src/inlay-hints.ts

```diff
@@ -0,0 +1,233 @@
+      const uri = editor.document.uri.toString();
+      const file = this.sourceFiles.get(uri) ?? {
+        document: editor.document,
+        cachedDecorations: null,
+        inlaysRequest: null
+      };
+      newSourceFiles.set(uri, file);
+
+      // No text documents changed, so we may try to use the cache
+      if (!file.cachedDecorations) {
+        const hints = await this.fetchHints(file);
+        if (!hints)
+          return;
+
+        file.cachedDecorations = this.hintsToDecorations(hints);
+      }
+
+      this.renderDecorations(editor, file.cachedDecorations);
+    });
+
+    // Cancel requests for no longer visible (disposed) source files
+    this.sourceFiles.forEach((file, uri) => {
+      if (!newSourceFiles.has(uri)) {
+        file.inlaysRequest?.cancel();
+      }
+    });
+
+    this.sourceFiles = newSourceFiles;
+  }
+
+  onDidChangeTextDocument({contentChanges,
+                           document}: vscode.TextDocumentChangeEvent) {
+    if (!this.enabled || contentChanges.length === 0 ||
+        !isClangdDocument(document))
+      return;
+    this.syncCacheAndRenderHints();
+  }
+
+  dispose() { this.stopShowingHints(); }
+
+  private startShowingHints() {
+    // Set up initial cache shape
+    this.context.visibleClangdEditors.forEach(
+        editor => this.sourceFiles.set(editor.document.uri.toString(), {
+          document: editor.document,
+          inlaysRequest: null,
+          cachedDecorations: null
+        }));
+
+    this.syncCacheAndRenderHints();
+  }
+
+  private stopShowingHints() {
+    this.sourceFiles.forEach(file => file.inlaysRequest?.cancel());
+    this.context.visibleClangdEditors.forEach(
+        editor =>
+            this.renderDecorations(editor, {beforeHints: [], afterHints: []}));
+  }
+
+  private renderDecorations(editor: vscode.TextEditor,
+                            decorations: InlayDecorations) {
+    editor.setDecorations(beforeHintStyle.decorationType,
+                          decorations.beforeHints);
+    editor.setDecorations(afterHintStyle.decorationType,
+                          decorations.afterHints);
+  }
+
+  private syncCacheAndRenderHints() {
+    this.sourceFiles.forEach(
+        (file, uri) => this.fetchHints(file).then(hints => {
+          if (!hints)
+            return;
+
+          file.cachedDecorations = this.hintsToDecorations(hints);
+
+          for (const editor of this.context.visibleClangdEditors) {
+            if (editor.document.uri.toString() == uri) {
+              this.renderDecorations(editor, file.cachedDecorations);
+            }
+          }
+        }));
+  }
+
+  private hintsToDecorations(hints: InlayHint[]): InlayDecorations {
+    const decorations: InlayDecorations = {beforeHints: [], afterHints: []};
+    const conv = this.context.client.protocol2CodeConverter;
+    for (const hint of hints) {
+      switch (hint.kind) {
+      case InlayHintKind.Parameter: {
```

Comment:
We don't (know how to) display hints of unknown types.
Could add a fallback (default) style, or leave it and make a note to ensure the LSP protocol doesn't have this problem.